Fix for #10: Hybrid Hash Join for iterations

rmetzger opened the pull request:
With this change, ozone's hash join implementation is able to be reopened multiple times with different probe side inputs.

In-memory build-site contents are spilled to disk once the build phase has finished. Spilled build partitions are kept until hybrid hash join's final close() call, so that subsequent probe phases can re-use the build side.
Upon a reopen() call, hybrid hash join will reinitialize itself to the same state as after a regular open() call, but without re-pulling the build side iterator.

If the build side fits completely in-memory, no spilling will happen.

I'm very happy about feedback and comments on my code!

rmetzger committed:
Implemented Hash Join so that is able to handle multiple probe phases

rmetzger committed:
Removed all my debug messages

rmetzger committed:
Some more cleanups

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/HashPartition.java:
```diff
@@ -345,15 +365,15 @@ public int finalizeProbePhase(List<MemorySegment> freeMemory, List<HashPartition
 			this.nextOverflowBucket = 0;
-			
 			// return the partition buffers
+			if(initialBuildSideChannel != null && !isRestored) return 0; // we already returned the partitionBuffers via the returnQueue.
```
please add a space between if and (. Please always use braces in conjunction with if

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/HashPartition.java:
```diff
@@ -415,6 +437,9 @@ public void clearAllMemory(List<MemorySegment> target)
 				this.probeSideChannel.close();
 				this.probeSideChannel.deleteChannel();
 			}
+			if(initialBuildSideChannel != null) {
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/HashPartition.java:
```diff
@@ -438,6 +463,99 @@ final SeekableDataOutputView getWriteView() {
+	
+	public void prepareProbePhase(IOManager ioAccess, Channel.Enumerator probeChannelEnumerator,
+            LinkedBlockingQueue<MemorySegment> bufferReturnQueue, List<Channel.ID> spilledPartitions) throws IOException {
+		if(isInMemory()) {
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/HashPartition.java:
```diff
@@ -438,6 +463,99 @@ final SeekableDataOutputView getWriteView() {
+			return 0;
+		}
+		
+		if(isInMemory()) {
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/HashPartition.java:
```diff
@@ -438,6 +463,99 @@ final SeekableDataOutputView getWriteView() {
+	 * @return Number of memorySegments in the writeBehindBuffers!
+	 */
+	int spillInMemoryPartition(Channel.ID targetChannel, IOManager ioManager, LinkedBlockingQueue<MemorySegment> writeBehindBuffers) throws IOException {
+		assert isInMemory();
```
use Preconditions.checkState() instead of assert

rmetzger replied:
I use assertion here more as a documentation. If nobody changes the code, this assertion will always be true. I hope that others that are changing the code have assertions enabled when changing this code.
It would be a waste of CPU cycles to use Preconditions here. Since assertions are a language construct, they are (hopefully) not evaluated when assertions are disabled.
If you absolutely don't want to have assertions in this code, I'd rather remove it. (because the method is only called if isInMemory() is true for the partition.

StephanEwen replied:
We have a policy to not use "assert". If you check the stratosphere Wiki, there are still code guidelines. Please have a look and try to follow them.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/HashPartition.java:
```diff
@@ -438,6 +463,99 @@ final SeekableDataOutputView getWriteView() {
+	void restorePartitionBuffers(IOManager ioManager, List<MemorySegment> availableMemory) throws IOException {
+		final BulkBlockChannelReader reader = ioManager.createBulkBlockChannelReader(this.initialBuildSideChannel, 
+			availableMemory, this.initialPartitionBuffersCount);
+		reader.close();
```
Sure you can close without a try/finally block here?

StephanEwen replied:
The logic for assured closing of files (in case of an abnormal termination of the algorithm) needs some reworking anyways, for the entire Hash Join. I would leave this for now.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MultiMatchMutableHashTable.java:
```diff
@@ -0,0 +1,159 @@
+	}
+
+	public void reopenProbe(MutableObjectIterator<PT> probeInput) throws IOException {
+		if(closed) {
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MultiMatchMutableHashTable.java:
```diff
@@ -0,0 +1,159 @@
+		// We restore the same "partitionsBeingBuild" state as after the initial open call.
+		partitionsBeingBuilt.addAll(initialPartitions);
+		
+		if(spilled) {
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MultiMatchMutableHashTable.java:
```diff
@@ -0,0 +1,159 @@
+			//setup partitions for insertion:
+			for (int i = 0; i < this.partitionsBeingBuilt.size(); i++) {
+				HashPartition<BT, PT> part = this.partitionsBeingBuilt.get(i);
+				if(part.isInMemory()) {
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MultiMatchMutableHashTable.java:
```diff
@@ -0,0 +1,159 @@
+				try {
+					p.prepareProbePhase(ioManager,currentEnumerator,writeBehindBuffers, null);
+				} catch (IOException e) {
+					e.printStackTrace();
```
rethrow or ignore and log, but never do printStrackTrace

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MultiMatchMutableHashTable.java:
```diff
@@ -0,0 +1,159 @@
+	 * @throws IOException 
+	 */
+	void storeInitialHashTable() throws IOException {
+		if(spilled) {
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MultiMatchMutableHashTable.java:
```diff
@@ -0,0 +1,159 @@
+	
+	@Override
+	public void close() {
+		if(partitionsBeingBuilt.size() == 0) { // partitions are cleared after the build phase. But we need to drop
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MutableHashTable.java:
```diff
@@ -473,20 +476,44 @@ public boolean nextRecord() throws IOException {
-
+		MultiMatchMutableHashTable<BT,PT> mmht = null;
+		boolean furtherPartitioning = false;
+		if(isMultiHashTable) {
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MutableHashTable.java:
```diff
@@ -473,20 +476,44 @@ public boolean nextRecord() throws IOException {
+					break;
+				}
+			}
+			if(furtherPartitioning) {
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MutableHashTable.java:
```diff
@@ -473,20 +476,44 @@ public boolean nextRecord() throws IOException {
-			HashPartition<BT, PT> p = this.partitionsBeingBuilt.get(i);
-			buffersAvailable += p.finalizeProbePhase(this.availableMemory, this.partitionsPending);
+			final HashPartition<BT, PT> p = this.partitionsBeingBuilt.get(i);
+			if(isMultiHashTable) {
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MutableHashTable.java:
```diff
@@ -473,20 +476,44 @@ public boolean nextRecord() throws IOException {
-		// release the table memory
-		releaseTable();
+		
+		if(!isMultiHashTable || furtherPartitioning || this.currentRecursionDepth > 0) { // if we are in an higher rec lvl, hashjoin is operating regularly
```
space between if and (.

sscdotopen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MutableHashTable.java:
```diff
@@ -734,7 +762,13 @@ protected void buildTableFromSpilledPartition(final HashPartition<BT, PT> p) thr
 				this.availableMemory, p.getBuildSideBlockCount());
-			reader.closeAndDelete(); // call waits until all is read
+			// call waits until all is read
+			if(isMultiHashTable && p.recursionLevel == 0) {
```
space between if and (. Sure you can close the reader without try/finally?

rmetzger replied:
This is the same reader as above. The code here is actually from Stephan. I only changed the statement so that we don't delete the file in some situations.

I'm not sure if I got your question right. I assume that it is good programming style to use the finally block to clean up the mess after an exception.
The only thing I could do if close() throws an exception is deleting the file hash join was about to read .. but HJ does not do this anywhere. So if Hashjoin fails, it will always leave some ".channel" files.
If the close() call fails, there is no chance for HJ to recover.

sscdotopen commented on pact/pact-runtime/src/test/java/eu/stratosphere/pact/runtime/hash/HashMatchMultiProbeTest.java:
```diff
@@ -0,0 +1,244 @@
+				Assert.fail("Collection for key " + entry.getKey() + " is not empty");
+		}
+		
+		for(int i = 0; i < NUM_PROBES; i++) {
```
space between for and (.

sscdotopen commented on pact/pact-runtime/src/test/java/eu/stratosphere/pact/runtime/hash/HashMatchMultiProbeTest.java:
```diff
@@ -0,0 +1,244 @@
+			
+			// assert that each expected match was seen for the second input
+			for (Entry<TestData.Key, Collection<RecordMatch>> entry : expectedNMatchesMapList.get(i).entrySet()) {
+				if (!entry.getValue().isEmpty())
```
always use braces with if

rmetzger committed:
github code review comments

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/services/memorymanager/spi/DefaultMemoryManager.java:
```diff
@@ -112,7 +112,7 @@ public DefaultMemoryManager(long memorySize, int pageSize)
 			throw new IllegalArgumentException("Size of total memory must be positive.");
 		}
-		if (pageSize <= MIN_PAGE_SIZE) {
+		if (pageSize <= MIN_PAGE_SIZE) { // FIXME "must be greater than X or the condition is wrong!
```
What is this comment about? Please remove it...

StephanEwen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/BuildFirstHashMatchIterator.java:
```diff
@@ -30,6 +30,7 @@
 import eu.stratosphere.pact.generic.types.TypePairComparator;
 import eu.stratosphere.pact.generic.types.TypeSerializer;
 import eu.stratosphere.pact.runtime.task.util.MatchTaskIterator;
+import eu.stratosphere.pact.runtime.test.util.TestData.GeneratorIterator;
```
Unnecessary import. Does that even build with maven, because it is a dependency on classes in the "test" scope.

Please activate warnings for unused imports and remove them from the code prior to checkin.

StephanEwen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/BuildFirstHashMatchIterator.java:
```diff
@@ -38,9 +39,9 @@
  *
  * @author Stephan Ewen
  */
-public final class BuildFirstHashMatchIterator<V1, V2, O> implements MatchTaskIterator<V1, V2, O>
```
Some classes are final to hint the compiler / Jitter to inline methods where possible. When removing teh final from the class, see whether you can make the critical methods final.

In this case, please make nextKey() final.

StephanEwen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/BuildFirstHashMultiMatchIterator.java:
```diff
@@ -0,0 +1,56 @@
+public class BuildFirstHashMultiMatchIterator<V1, V2, O> extends BuildFirstHashMatchIterator<V1, V2, O> {
+
+	
+	final private MultiMatchMutableHashTable<V1, V2> multiHashTable;
```
Use Java conventions for modifier order:

(visibility) (static) (final) variable

StephanEwen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MutableHashTable.java:
```diff
@@ -146,11 +146,11 @@
 	// -------------------------- Bucket Size and Structure -------------------------------------
 	
-	private static final int NUM_INTRA_BUCKET_BITS = 7;
+	static final int NUM_INTRA_BUCKET_BITS = 7;
```
Please use protected modifier, if the variable is intended for use in sub-classes (which are not necessarily in the same package).

StephanEwen commented on pact/pact-runtime/src/test/java/eu/stratosphere/pact/runtime/hash/HashMatchMultiProbeTest.java:
```diff
@@ -0,0 +1,246 @@
+ * Notes:
+ * 	- See AbstractCachedBuildSideMatchDriver
+ * 
+ * @author Robert Metzger
```
We are trying to get rid of @author tags to avoid that people feel "territorial". Feel free to also remove my and other people's @author tags when you come across them (we have not removed them in a coordinated fashion, I simply drop them when I come across them while editing a file).

Please add a file called "contributors" to the root directory and add your name there. By that it becomes more visible who was involved and at the same time, we avoid territorial behavior.

StephanEwen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/HashPartition.java:
```diff
@@ -476,7 +595,7 @@ protected int getLimitForSegment(MemorySegment segment) {
 	// ============================================================================================
 	
-	private static final class BuildSideBuffer<BT> extends AbstractPagedOutputView
+	static final class BuildSideBuffer<BT> extends AbstractPagedOutputView
```
Please use protected for methods relevant in sub-classes, rather than in other classes of the same package.

StephanEwen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MutableHashTable.java:
```diff
@@ -668,7 +695,7 @@ protected void buildInitialTable(final MutableObjectIterator<BT> input)
 		final int partitionFanOut = getPartitioningFanOutNoEstimates(this.availableMemory.size());
 		if (partitionFanOut > MAX_NUM_PARTITIONS) {
-			throw new RuntimeException("Hash join created ");
+			throw new RuntimeException("Hash join created "); //FIXME
```
What needs fixing here?

StephanEwen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/MutableHashTable.java:
```diff
@@ -470,20 +473,44 @@ public boolean nextRecord() throws IOException {
+		
 		// -------------- partition done ---------------
-
+		MultiMatchMutableHashTable<BT,PT> mmht = null;
```
Can you rework this a bit? Having a subclass for re-openable hash tables and then checking in the super-class whether we have actually the subclass seems like bad design.

StephanEwen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/BuildFirstHashMultiMatchIterator.java:
```diff
@@ -0,0 +1,56 @@
+import eu.stratosphere.pact.generic.types.TypePairComparator;
+import eu.stratosphere.pact.generic.types.TypeSerializer;
+
+public class BuildFirstHashMultiMatchIterator<V1, V2, O> extends BuildFirstHashMatchIterator<V1, V2, O> {
```
MultiMatch is a misleading name and reminds rather of a MultiMap (multiple values for the same key, which we have anyways).

Please rename the classes/variables to ReOpenableHashTable MultiStepHashTable or PersistentHashTable.

StephanEwen commented on pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/hash/HashPartition.java:
```diff
@@ -96,8 +97,25 @@
 	
 	private BlockChannelWriter probeSideChannel;		// the channel writer from the probe side, if partition is spilled
 	
+	// ------------------------------------------ Restoring ----------------------------------------------
```
Is it possible to move the changes to the Hash Partition into a subclass (re-openable partition, or so?)

rmetzger committed:
Better separation between Reopenable* and regular hash map using object orientation; all tests passed

rmetzger commented:
I addressed all issues from the comments here.
"mvn verify" run through without any errors.

Please have a look at the CONTRIBUTORS file and update your infos.

StephanEwen commented:
rebased and merged in e62e40230e7569420ca654106bdbe7d39e6920b1
